src/Opaleye/Internal/RunQuery.hs

```diff
@@ -344,6 +386,50 @@ jsonFieldTypeParser jsonTypeName field mData = do
         Just bs -> pure $ IPT.strictDecodeUtf8 bs
         _       -> returnError UnexpectedNull field ""
 
+jsonFieldTypeTextParser :: SBS.ByteString -> FieldParser ST.Text
+jsonFieldTypeTextParser jsonTypeName field mData = do
+    ti <- typeInfo field
+    if TI.typname ti == jsonTypeName
+       then convert
+       else returnError Incompatible field "types incompatible"
+  where
+    convert = case mData of
+        Just bs -> pure $ STE.decodeUtf8 bs
+        _       -> returnError UnexpectedNull field ""
+
+jsonFieldTypeLazyTextParser :: SBS.ByteString -> FieldParser LT.Text
+jsonFieldTypeLazyTextParser jsonTypeName field mData = do
+    ti <- typeInfo field
+    if TI.typname ti == jsonTypeName
+       then convert
+       else returnError Incompatible field "types incompatible"
+  where
+    convert = case mData of
+        Just bs -> pure . LTE.decodeUtf8 $ LBS.fromStrict bs
+        _       -> returnError UnexpectedNull field ""
+
+jsonFieldTypeByteParser :: SBS.ByteString -> FieldParser SBS.ByteString
+jsonFieldTypeByteParser jsonTypeName field mData = do
+    ti <- typeInfo field
+    if TI.typname ti == jsonTypeName
+       then convert
+       else returnError Incompatible field "types incompatible"
+  where
+    convert = case mData of
+        Just bs -> pure bs
+        _       -> returnError UnexpectedNull field ""
+
+jsonFieldTypeLazyByteParser :: SBS.ByteString -> FieldParser LBS.ByteString
+jsonFieldTypeLazyByteParser jsonTypeName field mData = do
+    ti <- typeInfo field
+    if TI.typname ti == jsonTypeName
+       then convert
+       else returnError Incompatible field "types incompatible"
+  where
+    convert = case mData of
+        Just bs -> pure $ LBS.fromStrict bs
+        _       -> returnError UnexpectedNull field ""
+
```

Comment:
Can you remove the duplication in here by doing all the others as an `(fmap . fmap) f jsonFieldTypeByteParser`? `f` will be `IPT.strictDecodeUtf8` for `String`, for example.

Reply:
I can 😛 . I was being lazy.

One moment.